### ISSUE TITLE
JIT should not force users to write code in ugly ways

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#48108]
Writing code for JIT is currently incredibly unpleasant, because it inherently wants you to split logic that belongs to the same function/module to different places. For example, take a look at the following two extremely common patterns:

1. Non-jittable parts of a function has to be broken into a separate function. (Or, equivalently, put the non-jittable parts into a python autograd fn, a workaround from @t-vi . )
    ```py
    def myfunc(x, y):
      ... input_check(x, y)
      z = _myfunc1(x, y)
      ... some non-jittable parts: e.g., io/error check
      return _myfunc2(x, y, z)
  
    @torch.jit.script
    def _myfunc1(x, y): ...
  
    @torch.jit.script
    def _myfunc2(x, y): ...
    ```

2. A scripted module has to break modularity. Users must **at creation time** script it, unlike the scripted functions. From the design perspective, sure you have to have a callable to script. However, from a UI perspective, users can never have modular code that works on both modules that should be jitted and modules that should not, unless they do the following hacks: 
    ```py
    class ModuleToScript(nn.Module):
        _should_script = True
       ...

    class ModuleNotToScript(nn.Module):
        _should_script = False
       ...

    def create_module(cls, ...):  # NB: users should *ONLY* use this to create module
        m = cls(...)
        if cls._should_script:
           m = torch.jit.script(m)
        return m
    ```

3. Tracing has similar issues.

The current JIT behavior is not "wrong" per se.  However, it really leads to bad coding patterns once you want to write some flexible code, to the extent that it feels actively working against you. And flexibility is really the thing that makes all us use PyTorch for research.

It also does not help that JIT UX improvement issues are almost never worked on: https://github.com/pytorch/pytorch/issues/29092, https://github.com/pytorch/pytorch/issues/29177, https://github.com/pytorch/pytorch/issues/16293, https://github.com/pytorch/pytorch/issues/40914

Related to 2: https://github.com/pytorch/pytorch/issues/25066

**Comment by suo** [comment]
@SsnL this is very helpful feedback. I think we have spent a lot of time over the last year writing new language features to unblock people, but we've not done a good job improving the fundamental usability of the system. My feeling is that the team is getting the feature side under control (plus we have more people now), so we can focus a lot more on usability issues going forward.

Do you think the issues you raised here are the most important usability issues overall? Or are there others captured in previous issues that you feel are more important to address first?

**Comment by t-vi** [comment]
I think yes, with the addition that 1. very much applies to classes, too, i.e. it'd be good to be able to script a module that has non-scripted submodules.

In the end, I think the UX we should be aiming at would be "is a JIT by default workflow reasonable" (i.e. all things that could be JITed will be JITed). It's probably not the default PyTorch behaviour anytime soon, but it should be possible to adopt that. (And with this, eliminate the "I need to JIT parts/my model" step in most cases.)

Independently, I've been revisiting making autograd.Function-like things in JIT, but I think that is actually less urgent.

**Comment by SsnL** [comment]
@suo Yeah, I think these are arguably more important (although some of the other issues may be easier to solve). While these may need nontrivial work. one good thing about solving these is that they drive more usage. In my experience (of using & talking with other pytorch users), the reason to not use JIT is usually not some missing language features, but how difficult it is to use. So when these are improved, more people will use JIT, and the pytorch team will also get more feedbacks and a better sense of what the general use cases / feature requests are like.

**Comment by SplitInfinity** [comment]
> Jittable parts of a (partially jittable) function has to be broken into a separate function. (Or, equivalently, put the non-jittable parts into a python autograd fn, a workaround from @t-vi . )

We do have `@torch.jit.ignore`, but it cannot be used as a context manager. If we added that functionality, we could do something like:
```
def fn(x, y):
  # code that is jittable
  with torch.jit.ignore():
    # code that is not jittable
  # code that is jittable
```

What do you think about this? It would unify `torch.jit.script.pure_python` that you proposed with the existing `torch.jit.ignore`.

> A scripted module has to break modularity. Users must at creation time script it, unlike the scripted functions. From the design perspective, sure you have to have a callable to script. However, from a UI perspective, users can never have modular code that works on both modules that should be jitted and modules that should not, unless they do the following hacks:

You can extend `ScriptModule` to create a class whose instances are always scripted:
```
     class Data(torch.jit.ScriptModule):
            @torch.jit.script_method
            def forward(self, x):
                return x.new_zeros(x.data.size())
```

**Comment by t-vi** [comment]
Personally, I would prefer to have a "generous mode" where the JIT will find what it cannot deal with. I'm currently looking into prototyping that, but my PR queue is rapidly filling with patches that conflict...

**Comment by SplitInfinity** [comment]
I think we talked about that on the Slack channel, I hope to try and prototype that via AST partitioning this weekend as we in the US have Monday off 🥳

**Comment by t-vi** [comment]
I look forward to it, too! You have a good long weekend!

**Comment by SsnL** [comment]
@SplitInfinity 
> You can extend ScriptModule to create a class whose instances are always scripted:
I don't consider that a valid option because the JIT team has been discouraging since 1.2.
```
 2. torch.jit.script(nn_module_instance) is now the preferred way to create ScriptModules, instead of inheriting from torch.jit.ScriptModule. These changes combine to provide a simpler, easier-to-use API for converting your nn.Modules into ScriptModules, ready to be optimized and executed in a non-Python environment.
 
...

As a result of these changes, the following items are considered deprecated and should not appear in new code:
The @torch.jit.script_method decorator

Classes that inherit from torch.jit.ScriptModule

The torch.jit.Attribute wrapper class

The __constants__ array

The torch.jit.annotate function
```

And also the `ScriptModule` doc formatting is terrible since at least 1.6.0 (https://pytorch.org/docs/stable/generated/torch.jit.ScriptModule.html#torch.jit.ScriptModule) and now on the doc site nowhere mentions inheritance as a valid option, except for the discouragement quoted above.

It is also unclear if that inheritance works with `@ignore` and `@export`.

**Comment by SsnL** [comment]
@SplitInfinity 
> What do you think about this? It would unify torch.jit.script.pure_python that you proposed with the existing torch.jit.ignore.

I don't have a particular favor for `pure_python` as a name, so it sounds good to me, except that `ignore` is a bit of a vague name.

**Comment by suo** [comment]
> I don't consider that a valid option because the JIT team has been discouraging since 1.2.

It was at some point discouraged because we felt the `torch.jit.script()` API was better, and we didn't want to incur the maintainability burden of having two ways to produce ScriptModules from nn.Modules. We have since actually done the work to make sure the underlying implementation is shared, so maintenance is no longer a concern. 

And since folks (including you guys) have pointed out there are valid use cases for the old API, I think it makes sense to stop officially discouraging it in any way.

We should sanity check that feature support hasn't drifted (e.g. you can do everything with the inheritance API as you can with the functional API) + update the docs to reflect that it's a valid option. @SsnL is your only concern with the inheritance API that we warn against it? Or are there functionality gaps we need to fix?

**Comment by SsnL** [comment]
@suo Yes, deprecation and lack of examples are my greatest concerns. It would be great to update the document to clearly show the options user have! Thanks for the quick response!

**Comment by tugsbayasgalan** [comment]
We are currently working on a project to use torch.jit.ignore as a context manager (https://github.com/pytorch/pytorch/pull/55172). There are still limitations that we need to figure out to make JIT figure out what it can't compile. But given the user annotate the input and output to the context manager, this feature would work. Will update this thread once it is merged. @suo @t-vi

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
